src/Xamarin.Android.Build.Tasks/Utilities/PreservePinvokesNativeAssemblyGenerator.cs

```diff
@@ -0,0 +1,406 @@
+				Align = 1, // it's a reference to C++ `bool`
+				Dereferenceable = 1,
+				IsCplusPlusReference = true,
+				NoCapture = true,
+				NonNull = true,
+				NoUndef = true,
+				WriteOnly = true,
+			},
+		};
+
+		var sig = new LlvmIrFunctionSignature (
+			name: "find_pinvoke",
+			returnType: typeof(IntPtr),
+			parameters: parameters,
+			new LlvmIrFunctionSignature.ReturnTypeAttributes {
+				NoUndef = true,
+			}
+		);
+
+		var func = new LlvmIrFunction (sig, MakeFindPinvokeAttributeSet (module)) {
+			CallingConvention = LlvmIrCallingConvention.Fastcc,
+			Visibility = LlvmIrVisibility.Hidden,
+		};
+		LlvmIrLocalVariable retval = func.CreateLocalVariable (typeof(IntPtr), "retval");
+		var state = new ConstructionState {
+			Func = func,
+			ReturnLabel = new LlvmIrFunctionLabelItem ("return"),
+			EntryPointHashParam = parameters[1],
+			Phi = new LlvmIrInstructions.Phi (retval),
+			Is64Bit = is64Bit,
+		};
+		module.Add (state.Func);
+		state.Func.Body.Add (new LlvmIrFunctionLabelItem ("entry"));
+
+		var libraryNameSwitchEpilog = new LlvmIrFunctionLabelItem ("libNameSW.epilog");
+		var componentSwitch = new LlvmIrInstructions.Switch<T> (parameters[0], libraryNameSwitchEpilog, "sw.libname");
+
+		state.Func.Body.Add (componentSwitch);
+		state.Phi.AddNode (libraryNameSwitchEpilog, null);
+
+		components.Sort ((Component a, Component b) => a.NameHash.CompareTo (b.NameHash));
+		Log.LogDebugMessage ("  Components to be preserved:");
+		uint componentID = 1;
+
+		foreach (Component component in components) {
+			Log.LogDebugMessage ($"    {component.Name} (hash: 0x{component.NameHash:x}; {component.PInvokes.Count} p/invoke(s))");
+
+			string comment = $" {component.Name} (p/invoke count: {component.PInvokes.Count})";
+			LlvmIrFunctionLabelItem componentLabel = AddSwitchItem<T> (componentSwitch, component.NameHash, is64Bit, comment, null);
+
+			func.Body.Add (componentLabel, comment);
+			AddPInvokeSwitch<T> (state, componentLabel, component, componentID++);
+		}
+
+		func.Body.Add (libraryNameSwitchEpilog);
+
+		var setKnownLib = new LlvmIrInstructions.Store (false, parameters[2]);
+		func.Body.Add (setKnownLib);
+		AddReturnBranch (func, state.ReturnLabel);
+
+		func.Body.Add (state.ReturnLabel);
+		func.Body.Add (state.Phi);
+		func.Body.Add (new LlvmIrInstructions.Ret (typeof (IntPtr), retval));
+	}
+
+	void AddPInvokeSwitch<T> (ConstructionState state, LlvmIrFunctionLabelItem componentLabel, Component component, uint id) where T: struct
+	{
+		var pinvokeSwitchEpilog = new LlvmIrFunctionLabelItem ($"pinvokeSW.epilog.{id}");
+		state.Phi.AddNode (pinvokeSwitchEpilog, null);
+
+		var pinvokeSwitch = new LlvmIrInstructions.Switch<T> (state.EntryPointHashParam, pinvokeSwitchEpilog, $"sw.pinvoke.{id}");
+		state.Func.Body.Add (pinvokeSwitch);
+
+		component.Sort ();
+		bool first = true;
+		foreach (PInvoke pi in component.PInvokes) {
+			string pinvokeName = pi.NativeFunction.Signature.Name;
+			string comment = $" {pinvokeName}";
```

Comment:
Looks like an implementation detail leaking. Can we just ensure that the comment is concatenated properly when is used rather than doing the space addition here?

Reply:
It's a byproduct of how the comments work in the LLVM IR generator. The generator assumes the caller knows how the comment should look and doesn't modify it at all. Without the space, the comment would be `;something` which I dislike. Making the LLVM IR generator inject the space before the comment interferes with the intention of the caller without the generator knowing it.